File: E-Commerce App Project (Tabbed)/Controllers/SingleItemViewController.m
//
//  SingleItemViewController.m
//  E-Commerce App Project (Tabbed)
//
//  Created by Rony Banik on 10/11/18.
//  Copyright © 2018 Rony Banik. All rights reserved.
//

#import "SingleItemViewController.h"
#import "ShoppingCart.h"

@interface SingleItemViewController (){
    ShoppingCart* checkoutCart;
}

@end

@implementation SingleItemViewController

@synthesize itemImage;
@synthesize itemName;
@synthesize itemCategory;
@synthesize itemID;
@synthesize itemPrice;
@synthesize itemBrand;
@synthesize itemQuality;


@synthesize setItemImage;
@synthesize setItemName;
@synthesize setItemCategory;
@synthesize setItemID;
@synthesize setItemPrice;
@synthesize setItemBrand;
@synthesize setItemQuality;

@synthesize itemObjectReceived;
@synthesize theItemObject;

@synthesize itemAlreadyAddedAlert;


- (void)viewDidLoad {
    [super viewDidLoad];
    checkoutCart = [ShoppingCart sharedInstance];
    self.addToCartStatusoutlet.selected = [checkoutCart containsItem:self.theItemObject] ? YES : NO;
    // Do any additional setup after loading the view.
    theItemObject = itemObjectReceived;
    itemImage.image=[UIImage imageWithData:setItemImage];
    itemName.text=setItemName;
    itemCategory.text=setItemCategory;
    itemID.text=setItemID.stringValue;
    itemPrice.text = [self showPrice:setItemPrice];
    itemBrand.text=setItemBrand;
    itemQuality.text=setItemQuality;
    
    itemAlreadyAddedAlert = [UIAlertController
                                 alertControllerWithTitle:@"Already Added To The Cart"
                                 message:@"Item Will Be Added To Your Cart Again"
                                 preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction* yesButton = [UIAlertAction
                                actionWithTitle:@"Got it!"
                                style:UIAlertActionStyleDefault
                                handler:^(UIAlertAction * action) {
                                    //Handle your yes please button action here
                                }];
    
    UIAlertAction* noButton = [UIAlertAction
                               actionWithTitle:@"Add To Cart Again!"
                               style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction * action) {
                                   ShoppingCart* checkoutCart = [ShoppingCart sharedInstance];
                                   [checkoutCart addItem:self.theItemObject];
                                   self.addToCartStatusoutlet.selected = YES;
                               }];
    
    [itemAlreadyAddedAlert addAction:yesButton];
    [itemAlreadyAddedAlert addAction:noButton];
    
}

- (void)viewWillAppear:(BOOL)animated{
    self.addToCartStatusoutlet.selected = [checkoutCart containsItem:self.theItemObject] ? YES : NO;
    if (!self.addToCartStatusoutlet.selected) {
        self.removeFromCartOutlet.hidden=YES;
    }else{
        self.removeFromCartOutlet.hidden=NO;
    }
}
- (void)viewDidAppear:(BOOL)animated{
    if (!self.addToCartStatusoutlet.selected) {
        self.removeFromCartOutlet.hidden=YES;
        [self.addToCartStatusoutlet setTitle:@"Add To Cart" forState:UIControlStateNormal];

    }else{
        self.removeFromCartOutlet.hidden=NO;
        [self.addToCartStatusoutlet setTitle:@"Add To Cart Again" forState:UIControlStateNormal];
    }
}
/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (IBAction)addToCartButton:(id)sender {
    
    if (!self.addToCartStatusoutlet.selected) {
        [checkoutCart addItem:self.theItemObject];
        self.addToCartStatusoutlet.selected = YES;
        self.removeFromCartOutlet.hidden=NO;
        [sender setTitle:@"Again Add To Cart" forState:UIControlStateNormal];
        [sender setBackgroundColor:[UIColor greenColor]];
        [sender setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    }else{
        self.removeFromCartOutlet.hidden=NO;
        [self presentViewController:itemAlreadyAddedAlert animated:YES completion:nil];
        [sender setTitle:@"Again Add To Cart" forState:UIControlStateNormal];
        [sender setBackgroundColor:[UIColor greenColor]];
        [sender setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    }
}

- (IBAction)removeFromCartButton:(id)sender {
    [checkoutCart removeItem:self.theItemObject];
    self.addToCartStatusoutlet.selected = NO;
    [sender setTitle:@"Add To Cart" forState:UIControlStateNormal];
    [sender setImage:[UIImage imageNamed:@"addToCart-icon-40.png"] forState:UIControlStateNormal];
    [sender setBackgroundColor:[UIColor blueColor]];
}

-(NSString *)showPrice:(NSNumber *)price{
    NSString *priceString = [NSString stringWithFormat:@"$%@",price];
    return priceString;
}
@end

File: E-Commerce App Project (Tabbed)/Controllers/SingleItemViewController.swift
//
//  SingleItemViewController.swift
//  E-Commerce App Project (Tabbed)
//
//  Converted to Swift with MVVM pattern
//

import UIKit

class SingleItemViewController: ViewController {
    
    // MARK: - ViewModel
    private let viewModel = SingleItemViewModel()
    
    // MARK: - IBOutlets
    @IBOutlet var itemImage: UIImageView!
    @IBOutlet var itemName: UILabel!
    @IBOutlet var itemCategory: UILabel!
    @IBOutlet var itemID: UILabel!
    @IBOutlet var itemPrice: UILabel!
    @IBOutlet var itemBrand: UILabel!
    @IBOutlet var itemQuality: UILabel!
    @IBOutlet var addToCartStatusoutlet: UIButton!
    @IBOutlet var removeFromCartOutlet: UIButton!
    
    var setItemImage: Data?
    var setItemName: String?
    var setItemCategory: String?
    var setItemID: NSNumber?
    var setItemPrice: NSNumber?
    var setItemBrand: String?
    var setItemQuality: String?
    
    var itemAlreadyAddedAlert: UIAlertController!
    var itemObjectReceived: Item?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if let item = itemObjectReceived {
            viewModel.configure(with: item)
        }
        
        setupUI()
        setupAlert()
        bindViewModel()
    }
    
    private func setupUI() {
        if let imageData = setItemImage {
            itemImage.image = UIImage(data: imageData)
        }
        itemName.text = setItemName
        itemCategory.text = setItemCategory
        itemID.text = setItemID?.stringValue
        itemPrice.text = "$\(setItemPrice ?? 0)"
        itemBrand.text = setItemBrand
        itemQuality.text = setItemQuality
    }
    
    private func setupAlert() {
        itemAlreadyAddedAlert = UIAlertController(
            title: "Already Added To The Cart",
            message: "Item Will Be Added To Your Cart Again",
            preferredStyle: .alert
        )
        
        let dismissAction = UIAlertAction(title: "Got it!", style: .default, handler: nil)
        
        let addAgainAction = UIAlertAction(title: "Add To Cart Again!", style: .default) { [weak self] _ in
            self?.viewModel.addToCart()
        }
        
        itemAlreadyAddedAlert.addAction(dismissAction)
        itemAlreadyAddedAlert.addAction(addAgainAction)
    }
    
    private func bindViewModel() {
        viewModel.isInCart.bind { [weak self] isInCart in
            guard let self = self else { return }
            self.addToCartStatusoutlet.isSelected = isInCart
            self.removeFromCartOutlet.isHidden = !isInCart
            self.addToCartStatusoutlet.setTitle(isInCart ? "Add To Cart Again" : "Add To Cart", for: .normal)
            if isInCart {
                self.addToCartStatusoutlet.backgroundColor = .green
                self.addToCartStatusoutlet.setTitleColor(.black, for: .normal)
            }
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        viewModel.refreshCartStatus()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        viewModel.refreshCartStatus()
    }
    
    @IBAction func addToCartButton(_ sender: UIButton) {
        if !viewModel.isInCart.value {
            viewModel.addToCart()
        } else {
            present(itemAlreadyAddedAlert, animated: true, completion: nil)
        }
    }
    
    @IBAction func removeFromCartButton(_ sender: UIButton) {
        viewModel.removeFromCart()
        sender.setTitle("Add To Cart", for: .normal)
        sender.setImage(UIImage(named: "addToCart-icon-40.png"), for: .normal)
        sender.backgroundColor = .blue
    }
}
